providers/default/kdfs/scrypt.c

```diff
@@ -13,268 +13,225 @@
+static OSSL_OP_kdf_derive_fn kdf_scrypt_derive;
+static OSSL_OP_kdf_settable_ctx_params_fn kdf_scrypt_settable_ctx_params;
+static OSSL_OP_kdf_set_ctx_params_fn kdf_scrypt_set_ctx_params;
+
 static int scrypt_alg(const char *pass, size_t passlen,
                       const unsigned char *salt, size_t saltlen,
                       uint64_t N, uint64_t r, uint64_t p, uint64_t maxmem,
-                      unsigned char *key, size_t keylen);
+                      unsigned char *key, size_t keylen, EVP_MD *sha256);
 
-struct evp_kdf_impl_st {
+typedef struct {
+    void *provctx;
     unsigned char *pass;
     size_t pass_len;
     unsigned char *salt;
     size_t salt_len;
     uint64_t N;
-    uint32_t r, p;
+    uint64_t r, p;
     uint64_t maxmem_bytes;
-};
-
-/* Custom uint64_t parser since we do not have strtoull */
-static int atou64(const char *nptr, uint64_t *result)
-{
-    uint64_t value = 0;
-
-    while (*nptr) {
-        unsigned int digit;
-        uint64_t new_value;
+    EVP_MD *sha256;
+} KDF_SCRYPT;
 
-        if ((*nptr < '0') || (*nptr > '9')) {
-            return 0;
-        }
-        digit = (unsigned int)(*nptr - '0');
-        new_value = (value * 10) + digit;
-        if ((new_value < digit) || ((new_value - digit) / 10 != value)) {
-            /* Overflow */
-            return 0;
-        }
-        value = new_value;
-        nptr++;
-    }
-    *result = value;
-    return 1;
-}
+static void kdf_scrypt_init(KDF_SCRYPT *ctx);
 
-static EVP_KDF_IMPL *kdf_scrypt_new(void)
+static void *kdf_scrypt_new(void *provctx)
 {
-    EVP_KDF_IMPL *impl;
+    KDF_SCRYPT *ctx;
 
-    impl = OPENSSL_zalloc(sizeof(*impl));
-    if (impl == NULL) {
-        KDFerr(KDF_F_KDF_SCRYPT_NEW, ERR_R_MALLOC_FAILURE);
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx == NULL) {
+        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+    ctx->provctx = provctx;
+    ctx->sha256 = EVP_MD_fetch(PROV_LIBRARY_CONTEXT_OF(provctx),
+                               "sha256", NULL);
+    if (ctx->sha256 == NULL) {
+        OPENSSL_free(ctx);
+        ERR_raise(ERR_LIB_PROV, PROV_R_UNABLE_TO_LOAD_SHA256);
         return NULL;
     }
-    kdf_scrypt_init(impl);
-    return impl;
+    kdf_scrypt_init(ctx);
+    return ctx;
 }
 
-static void kdf_scrypt_free(EVP_KDF_IMPL *impl)
+static void kdf_scrypt_free(void *vctx)
 {
-    kdf_scrypt_reset(impl);
-    OPENSSL_free(impl);
+    KDF_SCRYPT *ctx = (KDF_SCRYPT *)vctx;
+
+    kdf_scrypt_reset(ctx);
+    EVP_MD_meth_free(ctx->sha256);
+    OPENSSL_free(ctx);
 }
 
-static void kdf_scrypt_reset(EVP_KDF_IMPL *impl)
+static void kdf_scrypt_reset(void *vctx)
 {
-    OPENSSL_free(impl->salt);
-    OPENSSL_clear_free(impl->pass, impl->pass_len);
-    memset(impl, 0, sizeof(*impl));
-    kdf_scrypt_init(impl);
+    KDF_SCRYPT *ctx = (KDF_SCRYPT *)vctx;
+
+    OPENSSL_free(ctx->salt);
+    OPENSSL_clear_free(ctx->pass, ctx->pass_len);
+    memset(ctx, 0, sizeof(*ctx));
```

Comment:
Here, you have a different problem.  Since `sha256` is chosen internally, this will clear the fetched algo, to never be fetched again (because you only do that in `kdf_scrypt_new()`), which leads to problems in subsequent derivation calls (ctx->md == NULL ==> [kaboom]).  The reasonable thing here is to remove the memset.  This is fine, because you're re-initializing the remaining fields in the `kdf_scrypt_init()` call anyway.

Reply:
Of course, you could also do like you do in pbkdf2.c, and re-fetch the algo in the init function